std/experimental/typecons.d

```diff
@@ -1074,3 +1074,258 @@ pure nothrow @system unittest
+        _ok = true;
+    }
+
+    /// Construct from unexpected value `unexpectedValue.`
+    this(Unexpected!E unexpectedValue) @trusted
+    {
+        // TODO reuse opAssign?
+        moveEmplace(unexpectedValue, _unexpectedValue);
+        _ok = false;
+    }
+
+    /// Assign from expected value `expectedValue.`
+    void opAssign()(auto ref T expectedValue) @trusted
+    {
+        // TODO is this ok?:
+        clear();
+        moveEmplace(expectedValue, _expectedValue);
+        _ok = true;
+    }
+
+    /// Assign from unexpected value `unexpectedValue.`
+    void opAssign(E unexpectedValue) @trusted
+    {
+        clear();
+        moveEmplace(unexpectedValue, _unexpectedValue);
+        _ok = false;
+    }
+
+    /// Clear (empty) contents.
+    private void clear() @trusted
+    {
+        release();
+        static if (isAddress!T)
+        {
+            _expectedValue = null;
+        }
+    }
+
+    /// Release any memory used to store contents.
+    private void release() @trusted
+    {
+        import std.traits : hasElaborateDestructor;
+        if (hasExpectedValue)
+        {
+            static if (!is(T == class))
+            {
+                static if (hasElaborateDestructor!T)
+                {
+                    destroy(_expectedValue);
+                }
+            }
+            _ok = false;
+        }
+        else
+        {
+            static if (!is(E == class))
+            {
+                static if (hasElaborateDestructor!E)
+                {
+                    destroy(_unexpectedValue);
+                }
+            }
+            destroy(_unexpectedValue);
+            // TODO change _ok?
+        }
+    }
+
+    /** Is `true` iff this has a expectedValue of type `T`. */
+    bool hasExpectedValue() const
+    {
+        return _ok;
+    }
+
+    import std.traits : CommonType;
+
+    /** Get current value if any or call function `elseWorkFun` with compatible return value.
+     *
+     * TODO is this anywhere near what we want?
+     */
+    CommonType!(T, typeof(elseWorkFun())) valueOr(alias elseWorkFun)() const
+    if (is(CommonType!(T, typeof(elseWorkFun()))))
+    {
+        if (hasExpectedValue)
+        {
+            return expectedValue;
+        }
+        else
+        {
+            return elseWorkFun(); // TODO is this correct
+        }
+    }
+
+    import std.functional : unaryFun;
+
+    /** If `this` is an expected value (of type `T`) apply `fun` on it and
+     * return result, otherwise return current unexpected value (of type `E`).
+     *
+     * See_Also: https://dlang.org/phobos/std_typecons.html#.apply
+     */
+    Expected!(typeof(unaryFun!fun(T.init)), E) apply(alias fun)() @trusted
+    {
+        alias fn = unaryFun!fun;
+        if (hasExpectedValue)
+        {
+            return typeof(return)(fn(_expectedValue));
+        }
+        else
+        {
+            return typeof(return)(Unexpected!E(_unexpectedValue));
+        }
+    }
```

Comment:
See above.

Reply:
I guess `apply` needs to be external aswell?

Reply:
Apologies for missing this. Yes, it does.